lib/Backend/Lower.cpp

```diff
@@ -24604,7 +24660,37 @@ Lowerer::InsertReturnThunkForRegion(Region* region, IR::LabelInstr* restoreLabel
         }
 
         IR::LabelOpnd * continuationAddr;
-        if (region->GetParent()->GetType() != RegionTypeRoot)
+        if (region->IsNonExceptingFinally())
```

Comment:
Add lots of comments and asserts to explain the mechanics here and to enforce your invariants